wdl-cli/src/analysis/source.rs

```diff
@@ -0,0 +1,142 @@
+//! Sources for a WDL documents used in analysis.
+
+use std::path::Path;
+use std::path::PathBuf;
+
+use anyhow::Result;
+use anyhow::anyhow;
+use anyhow::bail;
+use path_clean::clean;
+use url::Url;
+use wdl_analysis::Analyzer;
+use wdl_engine::path::parse_url;
+
+/// A source for an analysis.
+#[derive(Clone, Debug)]
+pub enum Source {
+    /// A remote URL.
+    Remote(Url),
+
+    /// A local file.
+    File(Url),
+
+    /// A local directory.
+    Directory(PathBuf),
+}
+
+impl Source {
+    /// Attempts to reference the source as a URL.
+    pub fn as_url(&self) -> Option<&Url> {
+        match self {
+            Source::Remote(url) | Source::File(url) => Some(url),
+            Source::Directory(_) => None,
+        }
+    }
+
+    /// Registers the source within an [`Analyzer`].
+    pub async fn register<T: Send + Clone + 'static>(
+        self,
+        analyzer: &mut Analyzer<T>,
+    ) -> Result<()> {
+        match self {
+            Source::Remote(url) | Source::File(url) => analyzer.add_document(url).await,
+            Source::Directory(path) => analyzer.add_directory(path).await,
+        }
+    }
+}
+
+impl std::fmt::Display for Source {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Source::Remote(url) => write!(f, "{url}"),
+            Source::File(url) => write!(f, "{url}"),
+            Source::Directory(path) => write!(f, "{path}", path = path.display()),
+        }
+    }
+}
+
+impl std::str::FromStr for Source {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Some(url) = parse_url(s) {
+            return Ok(Self::Remote(url));
+        }
+
+        let path = Path::new(s);
+
+        let path = std::path::absolute(path)
+            .map_err(|_| anyhow!("failed to convert `{path}` to a URI", path = path.display()))
```

Comment:
```suggestion
            .map_err(|_| anyhow!("failed to make `{path}` absolute", path = path.display()))
```

Reply:
I actually did consider this one, and I think I like the more generic `failed to cover to a URI`. The reason for this is because, when a path is not able to be made absolute, there's probably something funkier going on (e.g., it's not a real path) in which case the error message is more concerning if it's that specific.